cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/InMemorySourceControlOperationRunner.java

```diff
@@ -66,12 +77,47 @@ public PushAppResponse push(NamespaceId namespace, RepositoryConfig repoConfig,
 
       //TODO: CDAP-20371, Add retry logic here in case the head at remote moved while we are doing push
       return writeAppDetailAndPush(repositoryManager, appToPush, commitDetails);
-    } finally {
-      try {
-        repositoryManager.close();
-      } catch (IOException e) {
-        LOG.warn("Failed to close the RepositoryManager, there may be leftover files", e);
+    }
+  }
+
+  @Override
+  public PullAppResponse pull(ApplicationReference appRef, RepositoryConfig repoConfig) throws PullFailureException,
+    NotFoundException, AuthenticationConfigException, IllegalArgumentException {
+    String applicationName = appRef.getApplication();
+    String configFileName = generateConfigFileName(applicationName);
+    LOG.info("Cloning remote to pull application {}", applicationName);
+    String contents;
+    String fileHash;
+    Path appRelativePath;
+    try (RepositoryManager repositoryManager = repoManagerFactory.create(appRef.getNamespaceId(), repoConfig)) {
+      appRelativePath = repositoryManager.getFileRelativePath(configFileName);
+      String commitID = repositoryManager.cloneRemote();
+      Path filePathToRead;
+      filePathToRead = validateAppConfigRelativePath(repositoryManager, appRelativePath);
+      if (!Files.exists(filePathToRead)) {
+        throw new NotFoundException(String.format("App with name %s not found in git repository", applicationName));
       }
+      contents = new String(Files.readAllBytes(filePathToRead), StandardCharsets.UTF_8);
+      LOG.info("Getting file hash for application {}", applicationName);
+      fileHash = repositoryManager.getFileHash(appRelativePath, commitID);
+    } catch (GitAPIException e) {
+      throw new PullFailureException(String.format("Failed to pull application %s: %s",
+                                                   applicationName,
+                                                   e.getMessage()), e);
+    } catch (Exception e) {
```

Comment:
Can you just do a separate multicatch with a single propagate?

Reply:
Also generally having such a multicatch is a bad smell. You should have a single parent exception for source control problems that you can catch all at once

Reply:
Created a multi catch block. Added a todo and opened https://cdap.atlassian.net/browse/CDAP-20410 to create new Exception classes.